### PR TITLE
[IPPInAppFeedback] Banner business logic and merchant eligibility in Orders screen

### PR DESCRIPTION
**Pull request by iamgabrielma**
Part of https://github.com/woocommerce/woocommerce-ios/issues/8530

## Description

This PR adds the logic to discern which in-app feedback banner we should show to the merchant, based on if the Store is eligible ( has COD enabled, and the country is IPP-supported ), and three conditions:

- Haven't collected any IPP payments.
- Have collected less than 10 IPP transactions in the last 30 days.
- Have collected more than 10 IPP transactions in the last 30 days.

At the moment we're not displaying or populating any specific content when any of the 3 conditions occur, this will be dealt with on a different PR.

## Changes
- Call to `displayIPPFeedbackBannerIfEligible()` in `OrderListViewController`'s init, which will perform the logic to check if COD is enabled, and fetch IPP transactions from storage (if any) as well as those within 30 days.

## Testing instructions
1. On a Store with IPP transactions, and COD enabled. Navigate to Menu > Orders
2. Check Xcode console and confirm that transactions within 30 days are logged:
```
hasResults? true
IPP transactions within 30 days: 3

[
  "OrderID: 1475 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-02 01:55:33 +0000)", 
  "OrderID: 1481 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-05 10:45:27 +0000)", 
  "OrderID: 1488 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-10 10:00:19 +0000)"
]

< 10 transactions within 30 days. Banner 2 shown
```

~~Due to the nature of fetching orders from Storage and how we use this layer as a sort of caching mechanism rather than as a source of truth, in some cases you may encounter 0 IPP transactions despite that these exist, for example if you install the app and go directly to the payments view without visiting the Orders view first. This is "expected" at this time, as we're looking for a fast solution to get us 90% there. The solution is to go to Orders > wait until are fetched > go back to Payments.~~

~~One potential way to smooth this problem is to add pull-to-refresh, this can be worked on a different PR or in i2 ( [tracked here](https://github.com/woocommerce/woocommerce-ios/issues/8530) )~~

Edit: As new requirements have changed the logic from Payments to Orders, this may no longer be a problem.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift [hunk 1]
```diff
@@ -62,9 +82,62 @@ final class InPersonPaymentsMenuViewModel {
...
+            let resultsCount = resultsController.fetchedObjects.count
+            let numberOfTransactions = 10
+
+            print("IPP transactions within 30 days: \(resultsCount)")
+            print(resultsController.fetchedObjects.map {
+                ("OrderID: \($0.orderID) - PaymentMethodID: \($0.paymentMethodID) - DatePaid: \(String(describing: $0.datePaid))")
+            })
+
+            if resultsCount == 0 {
+                // TODO: This option should use a different results controller or predicate, as we're looking for 0 orders historically, not within 30 days.
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift]: I'm wondering what would be the best approach here, as the 1st condition is "Stores with COD enabled, but no transactions" we're looking for 0 transactions historically, not within 30 days. 

Should we use two different results controller, or dynamically modify the predicate somehow? I'd think that if we fetch once to see if there's 0 or any orders, we could use that same results controller to filter down the latest 30 days without the need to fetch again.

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift [hunk 2]
```diff
@@ -62,9 +82,62 @@ final class InPersonPaymentsMenuViewModel {
...
+    /// Results controller that fetches IPP transactions within the last 30 days
+    ///
+    func createRecentIPPOrdersResultsController() -> ResultsController<StorageOrder> {
+        let today = Date()
+        let paymentGateway = Constants.wcpay
+        let thirtyDaysBeforeToday = Calendar.current.date(
+            byAdding: .day,
+            value: -30,
+            to: today
+        )!
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift]: I needed to force-unwrap this variable as the signature returns a `Date?`, however, I can't think of an alternative that we could use here as a default value if we happen to have a date that could not be calculated with the given input (which I don't see happening neither). One alternative could be to just not pass the `datePaid` and fetch everything in the predicate?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift]: Added a 2nd resultsController in https://github.com/woocommerce/woocommerce-ios/pull/8549/commits/218ba06237ee33f4d365a1ef20182b50e4e308ce

**Comment by iamgabrielma** [comment]
Thanks for the pairing session yesterday @joshheald , this PR is ready for review 🙇 

As discussed, in https://github.com/woocommerce/woocommerce-ios/pull/8549/commits/fe967c76c30c1ac9787ebd87cc1828ca4f684eb0 I've added an extra check to be sure that the IPP transactions we're getting are coming from the app, and not the web, by checking if their metadata contains a `receipts_url` key. 

I couldn't find an easy way to do so directly via the resultsController predicate, so I opted for [filtering the fetched objects once we get back the results from Storage](https://github.com/woocommerce/woocommerce-ios/blob/274766fe062d49656d7ac51ed354dff0e2f80cd7/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/InPersonPaymentsMenuViewModel.swift#L99-L106).

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsMenuViewModel.swift]: In https://github.com/woocommerce/woocommerce-ios/pull/8549/commits/0394db4cb5507639019b3b3b84cb8859cf5020fa I removed the force-unwrap in favour of providing a default Date value, better to get no results than to crash.

**Comment by joshheald** [comment]
@iamgabrielma Initial testing: all of my IPP stores say this –

```
COD enabled? true - Eligible Country? true
hasResults? false
IPP transactions within 30 days: 0
[]
0 transactions. Banner 1 shown

It was not possible to sync local orders settings: noOrdersSettings
```

However... some of those stores are full of recent IPP transactions, and all have at least some. I pulled to refresh and it says the same thing.

We can pair on why it's doing that if you like?

**Comment by joshheald** [comment]
Hmm. I had to launch it again to hit a breakpoint (because I wanted to hit the one in `viewDidLoad`) and the second time round, it came up with

```
COD enabled? true - Eligible Country? true
hasResults? true
IPP transactions within 30 days: 10
["OrderID: 440 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-11 18:49:54 +0000)", "OrderID: 445 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-13 14:24:28 +0000)", "OrderID: 450 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-16 10:50:36 +0000)", "OrderID: 451 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-13 17:40:57 +0000)", "OrderID: 452 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-13 16:48:11 +0000)", "OrderID: 455 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-13 17:45:11 +0000)", "OrderID: 456 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-16 10:49:15 +0000)", "OrderID: 457 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-16 10:37:57 +0000)", "OrderID: 458 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-16 10:22:59 +0000)", "OrderID: 459 - PaymentMethodID: woocommerce_payments (WooCommerce In-Person Payments - DatePaid: Optional(2023-01-16 11:06:49 +0000)"]
>= 10 transactions within 30 days. Banner 3 shown
```

So that's correct, there are exactly 10 paid IPP transactions in the last 30 days.

I think the fix is probably to refresh the results on viewWillAppear, or when we fetch orders, or something like that.

**Review by joshheald — APPROVED** [review]
Second time round it worked for me: I suspect there's something about the first time run for each site which means it's not quite working right. Worth investigating but can be done as follow on work if you want to merge this – it's feature flagged.

Consider adding tests for this logic too: it could help you work faster on it, even though this is a temporary addition to the app.

**Comment by iamgabrielma** [comment]
Thanks for the review!

> some of those stores are full of recent IPP transactions, and all have at least some. I pulled to refresh and it says the same thing.

That's interesting, from these 2 lines would seem that not only failed to fetch IPP orders but also to sync stored Orders settings:
```
hasResults? false
It was not possible to sync local orders settings: noOrdersSettings
```
I've [opened a new issue here](https://github.com/woocommerce/woocommerce-ios/issues/8654) to investigate this further, happy to pair at any time 🙇 

> Consider adding tests for this logic too: it could help you work faster on it, even though this is a temporary addition to the app.

[Will do](https://github.com/woocommerce/woocommerce-ios/issues/8655)!

**Pull request merged**